runtimes/ruby-3.1/server.rb

```diff
@@ -3,117 +3,255 @@
+  body = body_string
+  method = request.request_method
+  headers = {}
 
-  begin
-    load(USER_CODE_PATH + '/' + ENV['INTERNAL_RUNTIME_ENTRYPOINT'])
-  rescue Exception => e
-    p e
-    status 500
-    return {stderr: e.backtrace.join("\n") }.to_json
+  if request.env['CONTENT_TYPE'] != nil
+    headers['content-type'] = request.env['CONTENT_TYPE']
+  end
+
+  if request.env['CONTENT_LENGTH'] != nil
+    headers['content-length'] = request.env['CONTENT_LENGTH']
+  end
+
+  request.env.each do |header, value|
+    if header.start_with?('HTTP_')
+      header = header[5..-1].gsub("_", "-").downcase
+
+      if !header.start_with?('x-open-runtimes-')
+        headers[header] = value
+      end
+    end
   end
 
-  unless defined?(main = ())
-    status 500
-    return { stderr: 'File does not specify a main() function.' }.to_json
+  content_type = request.env['CONTENT_TYPE']
+  content_type = 'text/plain' if content_type.nil?
+
+  if content_type.include?('application/json')
+    if !(body_string.empty?)
+      body = JSON.parse(body_string)
+    end
   end
 
+  context_req = RuntimeRequest.new(url, method, scheme, host, port, path, query, query_string, headers, body, body_string)
+  context_res = RuntimeResponse.new
+  context = Context.new(context_req, context_res)
+
+  customstd = nil
+  
+  output = nil
+
   begin
+    load(USER_CODE_PATH + '/' + ENV['OPEN_RUNTIMES_ENTRYPOINT'])
+
+    unless defined?(main = ())
+      raise 'User function is not valid.'
+    end
+
     system_out = $stdout
     system_err = $stderr
-    user_out = StringIO.new
-    user_err = StringIO.new
-    $stdout = user_out
-    $stderr = user_err
+    customstd = StringIO.new
+    $stdout = customstd
+    $stderr = customstd
 
-    user_response = main(requestData, runtimeResponse)
-
-    response = {
-      response: user_response,
-      stdout: user_out.string
-    }.to_json
+    output = main(context)
   rescue Exception => e
-    status 500
-    return {
-      stdout: user_out.string,
-      stderr: user_err.string + "\n" + e.backtrace.join("\n")
-    }.to_json
+    context.error(e)
+    context.error(e.backtrace.join("\n"))
+    output = context.res.send('', 500, {})
   ensure
     $stdout = system_out
     $stderr = system_err
-    user_out = nil
-    user_err = nil
   end
 
-  status 200
-  return response
+  if output == nil
+    context.error('Return statement missing. return context.res.empty() if no response is expected.')
+    output = context.res.send('', 500, {})
+  end
+
+
+  output['body'] = '' if output['body'].nil?
+  output['statusCode'] = 200 if output['statusCode'].nil?
+  output['headers'] = {} if output['headers'].nil?
+
+  output['headers'].each do |header, value|
+    if !header.downcase.start_with?('x-open-runtimes-')
+      response.headers[header.downcase] = value
+    end
+  end
+
+  if customstd.string != nil && !customstd.string.empty?
+    context.log('Unsupported log noticed. Use context.log() or context.error() for logging.')
```

Comment:
Is "noticed" the best wording here? Wouldn't "detected" work better?

Reply:
Using detected instead.